File: balanced_brackets/stacks_balanced_brackets.py
class Stack(object):
    def __init__(self):
        self.data = []
    
    def push(self, item):
        self.data.append(item)
        
    def pop(self):
        return self.data.pop()
        
    def top(self):
        """ Don't forget to handle the error when there's no data """
        if len(self.data) == 0:
            return None
        else:
            return self.data[-1]
        
    def empty(self):
        return len(self.data) == 0
        
def is_matched(brackets):
    bracket_matches = {
        '}': '{',
        ']': '[',
        ')': '('
    }
    check_matches = Stack()
    for bracket in brackets:
        if bracket in bracket_matches.keys():
            if check_matches.top() == bracket_matches[bracket]:
                check_matches.pop()
            else:
                return False
        else:
            check_matches.push(bracket)
    if check_matches.empty():
        return True
    else:
        return False

t = int(input().strip())
for a0 in range(t):
    expression = input().strip()
    if is_matched(expression) == True:
        print("YES")
    else:
        print("NO")

File: balanced_brackets/stacks_balanced_brackets.js
function main() {
    function isBracket(brackets) {
        bracketMatches = {
            '}': '{',
            ')': '(',
            ']': '['
        };
        var checkMatches = [];
        for (var i = 0; i < brackets.length; i++) {
            bracket = brackets.charAt(i);
            if (bracket in bracketMatches) {
                if (bracketMatches[bracket] == checkMatches[checkMatches.length - 1]) {
                    checkMatches.pop();
                } else {
                    return false;
                }
            } else {
                checkMatches.push(bracket);
            }
        }
        if (checkMatches.length == 0) {
            return true;
        } else {
            return false;
        }
    }
    
    var t = parseInt(readLine());
    for(var a0 = 0; a0 < t; a0++){
        var expression = readLine();
        if (isBracket(expression)) {
            console.log("YES");
        } else {
            console.log("NO")
        }
    }
}